pepper-apis/src/main/java/org/broadinstitute/ddp/event/publish/pubsub/EventActionPubSubPublisher.java

```diff
@@ -0,0 +1,120 @@
+package org.broadinstitute.ddp.event.publish.pubsub;
+
+import static java.lang.String.format;
+import static org.broadinstitute.ddp.event.pubsubtask.api.PubSubTask.ATTR_TASK_TYPE;
+
+import java.io.IOException;
+import java.util.Map;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutureCallback;
+import com.google.api.core.ApiFutures;
+import com.google.api.gax.rpc.ApiException;
+import com.google.cloud.pubsub.v1.Publisher;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.protobuf.ByteString;
+import com.google.pubsub.v1.ProjectTopicName;
+import com.google.pubsub.v1.PubsubMessage;
+import com.typesafe.config.Config;
+import org.broadinstitute.ddp.constants.ConfigFile;
+import org.broadinstitute.ddp.event.publish.EventActionPublisher;
+import org.broadinstitute.ddp.housekeeping.PubSubConnectionManager;
+import org.broadinstitute.ddp.model.activity.types.EventActionType;
+import org.broadinstitute.ddp.util.ConfigManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implementation of {@link EventActionPublisher} providing publishing
+ * an event of a specified type to Google PubSub topic.<br>
+ * The parameters for PubSub publishing are taken from config.
+ * <b>Event publishing config parameters</b>:
+ * <ul>
+ *     <li>`usePubSubEmulator` - (true or false)</li>
+ *     <li>`googleProjectId` - project ID (for a pubsub topic)</li>
+ *     <li>`pubsub.eventActionTopic` - name of PubSub topic where to publish event</li>
+ * </ul>
+ */
+public class EventActionPubSubPublisher implements EventActionPublisher {
+
+    private static final Logger LOG = LoggerFactory.getLogger(EventActionPubSubPublisher.class);
+
+    public static final String ATTR_STUDY_GUID = "studyGuid";
+    public static final String ATTR_PARTICIPANT_GUID = "participantGuid";
+
+    private final Config cfg;
+    private final PubSubConnectionManager pubsubConnectionManager;
+    private final Publisher eventPublisher;
+    private final ProjectTopicName pubSubTopicName;
+
+
+    /**
+     * Constructor. Creates PubSub topic for publishing events.
+     */
+    public EventActionPubSubPublisher() throws IOException {
+        this.cfg = ConfigManager.getInstance().getConfig();
+        final boolean usePubSubEmulator = cfg.getBoolean(ConfigFile.USE_PUBSUB_EMULATOR);
+        this.pubsubConnectionManager = new PubSubConnectionManager(usePubSubEmulator);
+        this.pubSubTopicName = ProjectTopicName.of(
+                cfg.getString(ConfigFile.GOOGLE_PROJECT_ID),
+                cfg.getString(ConfigFile.PUBSUB_EVENT_ACTION_TOPIC));
+        this.eventPublisher = pubsubConnectionManager.getOrCreatePublisher(pubSubTopicName);
```

Comment:
These feel like pretty heavy operations... should we have a singleton "pubsub task" publisher that we can share?

Reply:
I agree about singleton. But I'd better not touch "pubsub task" for a while. As anyway we can not reuse this service in this (having opposite purpose) task.
May be we could refactor "pubsub task" later when implement such publisher singleton and reuse it in PubSub task service. The matter is that PubSub service initialized in the Housekeeping class where it is already initialized some of PubSub classes.. But in this current task (to process events and send them to DSM via PubSub) we have a problem to pass parameters from main Housekeeping class because `UpdateCustomWorkflowEventAction.doAction()` is called from `EventConfiguration.doAction()` which eventually called by `EventService.processAllActionsForEventSignal` which triggered in different parts of DSS code.. Therefore we need somehow to read configuration/initialize pubsub publisher, we can not reuse classes already instantiated in class `Housekeeping`. But anyway I agree about the moving PubSub initialization code to a separate singleton.
Doing this task we need somehow to keep in mind that similar code we have in `Housekeeping` and it needs to create common solution for entire DSS..

Reply:
PubSub publisher initialization code is moved out from the `EventActionPubSubPublisher.java` to class `PubSubPublisherInitializer.java`.